### ISSUE TITLE
Add support for more StringSyntaxAttribute syntaxes

### ISSUE THREAD

**Issue by stephentoub** [dotnet/roslyn#60593]
.NET 7 adds the new [StringSyntax(...)] attribute, which Roslyn is now already recognizing in support of the StringSyntaxAttribute.DateTimeFormat/Regex/Json values.  https://github.com/dotnet/runtime/pull/67621 is rounding out the named syntaxes on the attribute with an additional set, while also attributing all relevant parameters across the core libraries with these:

- [ ] "CompositeFormat" - a composite format string (https://docs.microsoft.com/en-us/dotnet/standard/base-types/composite-formatting#composite-format-string), like the one passed to string.Format / StringBuilder.AppendFormat.  It'd be great if syntax coloring could be applied to these, as is done for interpolated strings, and ideally syntax validation for composite format strings with invalid syntax (the latter could also be done as an analyzer).
- [ ] "DateOnlyFormat" / "TimeOnlyFormat" - same as "DateTimeFormat", but with only the date-related and time-related specifiers, relatively.  It'd be great if similar IntelliSense could be provided for these as is provided for "DateTimeFormat", just with only the relevant items filtered.
- [ ] "EnumFormat" - a specifier used for how an enum is formatted (https://docs.microsoft.com/en-us/dotnet/standard/base-types/enumeration-format-strings).  IntelliSense could do something similar to that for "DateTimeFormat", and warn when an unsupported option is used.
- [ ] "GuidFormat" - a specifier used for how a Guid is formatted or parsed (https://docs.microsoft.com/en-us/dotnet/api/system.guid.system-iformattable-tostring?msclkid=b56bc8efb52c11ecbb60280f9226ae5d&view=netstandard-1.6#parameters).   IntelliSense could do something similar to that for "DateTimeFormat", and warn when an unsupported option is used.
- [ ] "NumericFormat" - a specifier used for how one of the many numerical types is formatted (https://docs.microsoft.com/en-us/dotnet/standard/base-types/standard-numeric-format-strings?msclkid=027a8ed2b52d11ecad542115310f6ce9). IntelliSense could do something similar to that for "DateTimeFormat".
- [ ] "TimeSpanFormat" - a specifier used for how to format TimeSpan instances (https://docs.microsoft.com/en-us/dotnet/standard/base-types/standard-timespan-format-strings)
- [ ] "Uri" - URLs. Syntax coloring and validation could be employed.
- [ ] "Xml" - XML. Syntax coloring and validation could be employed.

cc: @CyrusNajmabadi

**Comment by CyrusNajmabadi** [comment]
I would be fine with support being added for all of these.  The hope of roslyn though would be that we don't own most of these though.  The purpose of my efforts to expose extensibility points here is so htat others can own this.

**Comment by Sergio0694** [comment]
Related (also cc. @teo-tsirpanis), would it be possible to also support syntax highlight for languages that the IDE already supports? Essentially what we get with the 3 backticks + language id on GitHub/Discord/etc., which could work whenever the language id (ie. the file extension) is one that VS already has support for.

For instance, if I had a `[StringSyntax("cpp")]` parameter (or `"cs"`, or whatever), as long as that id was one which would give me syntax highlight if I was to just open a file with that extension in VS, I would expect the same to apply to the literal as well.

Example scenario: I have a `[D2DPixelShaderSource]` attribute in [ComputeSharp](https://github.com/Sergio0694/ComputeSharp) you can use like this:

![image](https://user-images.githubusercontent.com/10199417/183922560-d966188f-d355-4db5-8dd1-db70210ddd1b.png)

The parameter of that `[D2DPixelShaderSource]` has `"hlsl"` as `[StringSyntax]` argument. Given I do get syntax highlight whenever I open an .hlsl file in VS, I would expect the same to apply here as well. Would this be doable? I mean VS evidently does have logic to do syntax highlight based on the file extension, but it be doable to plug that through here as well? 🙂

**Comment by CyrusNajmabadi** [comment]
Our intent is to expose this for plugins to be written.

**Comment by Sergio0694** [comment]
Right, but I guess what I'm wondering is: if the identifier is an extension that VS already has support for, couldn't this just work out of the box just like when you open a file, without the need for consumers to install any plugins or really anything at all? 😄
It would really be convenient especially when using identifiers that are widely used and with built-in support in VS.

**Comment by teo-tsirpanis** [comment]
@CyrusNajmabadi by "plugins" you mean Visual Studio extensions? What I would like is to allow library authors to define a custom language syntax with possible tooltips within the assembly, for example with an embedded resource file that gets read by IDEs. It would enable writing custom languages for both VS and VSCode. And of course also recognize the languages VS already knows as @Sergio0694 said.

**Comment by CyrusNajmabadi** [comment]
> Right, but I guess what I'm wondering is: if the identifier is an extension that VS already has support for, couldn't this just work out of the box just like when you open a file

Anything is possible. A core difficulty is that these are not the same languages. For example , in a c# string you have escape sequences that the underlying language will not understand.

**Comment by CyrusNajmabadi** [comment]
> It would really be convenient especially when using identifiers that are widely used and with built-in support in VS.

Sure. But someone has to pay for all that work. :-)

For context, json and regex was multiple weeks of work to try to get those working. Now someone has to test and validate every language embedded in c#

**Comment by rauhs** [comment]
Just an idea (feel free to ignore). Should something like this be possible:

```csharp
[StringSyntax("lang_de")]

[StringSyntax("lang_en-us")]
```

so IDEs can offer context dependent spell checking?

**Comment by teo-tsirpanis** [comment]
I don't think it would be of much value. Localized strings should be in a resource file either way.

**Comment by gdifilippo-ca** [comment]
Hi, I'm working on a free Visual Studio extension that adds support for GraphQL syntax. The extension is called [GraphQLTools](https://marketplace.visualstudio.com/items?itemName=codearchitects-research.GraphQLTools) and it is currently in preview. The source code is available [here](https://github.com/codearchitects/GraphQLTools); I hope that it could be of help to people interested in extending the feature to more languages.
Writing a Visual Studio extension is quite challenging because the documentation is not that exhaustive and we would love to be able to write plugins for the StringSyntaxAttribute feature in a more standardized way.

**Comment by Meir017** [comment]
This can we useful for playwright-dotnet and puppeteer-sharp for having literal JavaScript & HTML & CSS strings.

using [puppeteer-sharp](https://github.com/hardkoded/puppeteer-sharp):
```cs
await using var page = await browser.NewPageAsync();
await page.SetContentAsync("<div>My Receipt</div>"); // this can be highlighted as HTML
var result = await page.GetContentAsync();
var watchDog = page.WaitForFunctionAsync("()=> window.innerWidth < 100"); // this can be highlighted as JavaScript
await page.WaitForSelectorAsync("div.main-content"); // this can be highlighted as CSS
```
FYI @kblok

using [playwright-dotnet](https://github.com/microsoft/playwright-dotnet)
```cs
using var playwright = await Playwright.CreateAsync();
await using var browser = await playwright.Chromium.LaunchAsync(new() { Headless = false });
var page = await browser.NewPageAsync();
var href = await page.EvaluateAsync<string>("document.location.href"); // // this can be highlighted as JavaScript
await page.SetContentAsync("<div>My Receipt</div>"); // this can be highlighted as HTML
```

FYI @pavelfeldman @mxschmitt

**Comment by raresChelariu** [comment]
We have support in Rider for this !! Check out the [LanguageInjectionAttribute](https://blog.jetbrains.com/dotnet/2023/01/12/language-injection-improvements-in-rider-2022-3/#better-with-jetbrains-annotations) available in Rider 2021.3 or higher ! You can also enable it with a comment like `/*lang=css*/` before a e.g. local const variable name.

**Comment by ds5678** [comment]
Are there any plans to enable third-party support for syntaxes? Ideally, analyzers could provide highlighting and diagnostics at the same time.

**Comment by CyrusNajmabadi** [comment]
There are no plans to expose this currently
